proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1085 @@
+
+Then ``new 57`` can be shared arbitrarily. This is a phenomenon known
+as “promotion”: expressions without linear free variables are
+unrestricted. The typical solution is for ``new`` to take a
+continuation as an argument
+
+::
+
+   new :: Int -> (MArray %1 -> Ur a) %1 -> Ur a
+
+This forces the array to be single-threaded (thanks to the ``Ur a``
+return type, the ``MArray`` cannot escape the continuation's scope),
+which we can use to guarantee uniqueness.
+
+This is a little clumsy to program with. But more importantly, these
+continuations aren't very composable as argued in `this blog post
+<blog_scopes_>`_. See also the long discussion at
+`tweag/linear-base#130
+<https://github.com/tweag/linear-base/issues/130>`_. This proposal
+will let us define ``new`` in direct style.
+
+Proposed Change Specification
+-----------------------------
+
+Main Syntax Changes
+^^^^^^^^^^^^^^^^^^^
+
+Currently, type class constraints in GHC do not support multiplicty
+annotations.  GHC currently defines the syntax for type signatures as:
+
+::
+
+   ctype   ::= context '=>' ctype | type | ...
+
+Essentially, type signatures can consist of (among other things which
+we ignore here) zero or more qualified type arrows ``=>`` followed by
+a type. Here ``context`` is a list of class constraints.
+
+When ``-XLinearTypes`` is enabled, the following new syntax is
+enabled:
+
+::
+
+   ctype ::= context '%' 1 '=>' ctype | context '=>' ctype | type | ...
+
+Note that unlike multiplicities for function type arrows, linear
+constraint arrows may only be instantiated with a ``1`` (linear)
+multiplicity. Unlike linear function types, this proposal does not
+introduce multiplicity polymorphism in constraint arrows, so there is
+never a need for the multiplicity to be anything other than a ``1`` -
+an ``Many`` multiplicity is already represented by omitting the
+multiplicity entirely (i.e. using a standard constraint). Standard
+non-linear constraints can then still be used in combination with
+linear ones, with the order of linear and unrestricted constraints not
+mattering.
+
+Semantic/Desugaring
+^^^^^^^^^^^^^^^^^^^
+
+As typical, we define the semantics of linear constraints via
+desugaring. The linear fat arrow ``%1 =>`` desugars to a linear arrow
+``%1 ->``. Namely, writing ``⦇·⦈`` for the desugaring function:
+
+- ``⦇C %1 => A⦈ = C %1 -> ⦇A⦈``
+- ``⦇e :: C %1 => A⦈ = \(%1 $d :: C) -> ⦇e :: A⦈``
+
+This desugaring means that changes to GHC Core itself are not
+required: we only need the material introduced by linear types and
+already implemented.
+
+
+Using type classes linearly
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type class methods require an unrestricted class constraint:
+
+::
+
+  -- Given
+  class Foo a where
+    f :: F a
+    g :: G a
+
+  -- We have
+  f :: Foo a => F a
+  g :: Foo a => G a
+
+This is unchanged. But we add one exception: in type classes with exactly
+one method, the one method is linear in the class constraint
```

Comment:
I don't really understand the motivation for this. I get a sense it is justified by the need to use the `Linearly` in a linear way but it seems rather vague to me. Perhaps it could be spelled out more fully.

Reply:
I suppose this needs an explanation somewhere. It's not particularly related to the `Linearly` type class. Rather, it's to address that we need at least one function of type `C %1 => T`, otherwise we couldn't in any circumstance have a term of type `C %1 => U` (they would fail either because the `C` is used in a function of type `C =>`, or not used at all, neither of which is permitted).